src/Http/Controllers/SocialiteController.php

```diff
@@ -0,0 +1,136 @@
+
+class SocialiteController extends Controller
+{
+    use ConnectsToSocialProvider;
+
+    /**
+     * The guard implementation.
+     *
+     * @var \Illuminate\Contracts\Auth\StatefulGuard
+     */
+    protected $guard;
+
+    /**
+     * The creates user implementation.
+     *
+     * @var  \Laravel\Jetstream\Contracts\CreatesUserFromProvider;
+     */
+    protected $createsUser;
+
+    /**
+     * Create a new controller instance.
+     *
+     * @param  \Illuminate\Contracts\Auth\StatefulGuard  $guard
+     * @return void
+     */
+    public function __construct(StatefulGuard $guard, CreatesUserFromProvider $createsUser)
+    {
+        $this->guard = $guard;
+        $this->createsUser = $createsUser;
+    }
+
+    /**
+     * Sets the user's password.
+     *
+     * @param  \Illuminate\Http\Request  $request
+     * @param  \Laravel\Jetstream\Contracts\SetsUserPasswords  $updater
+     * @return \Illuminate\Http\Response
+     */
+    public function setPassword(Request $request, SetsUserPasswords $setter)
+    {
+        $setter->set($request->user(), $request->all());
+
+        return $request->wantsJson()
+                    ? new JsonResponse('', 200)
+                    : back()->with('status', 'password-set');
+    }
+
+    /**
+     * Get the redirect for the given Socialite provider.
+     *
+     * @param  \Illuminate\Http\Request  $request
+     * @param  string  $provider
+     * @return \Symfony\Component\HttpFoundation\RedirectResponse
+     */
+    public function redirectToProvider(Request $request, string $provider)
+    {
+        return Socialite::driver($provider)->redirect();
+    }
+
+    /**
+     * Attempt to log the user in via the provider user returned from Socialite.
+     *
+     * @param  \Illuminate\Http\Request  $request
+     * @param  string  $provider
+     * @return \Illuminate\Routing\Pipeline
+     */
+    public function handleProviderCallback(Request $request, string $provider)
+    {
+        if ($request->has('error')) {
+            return Auth::check()
+                ? redirect(config('fortify.home'))->dangerBanner($request->error_description)
+                : redirect()->route('register')->withErrors($request->error_description);
+        }
+
+        $providerUser = Socialite::driver($provider)->user();
+
+        $existing = ConnectedAccount::firstWhere([
+            'provider_id' => $providerUser->getId(),
+            'provider_name' => $provider,
+        ]);
+
+        if ($existing && Auth::check() && Auth::user()->id != $existing->user_id) {
+            return redirect(config('fortify.home'))->dangerBanner(
+                __('That account is already associated with a user. Please use a different account.')
+            );
+        }
+
+        $user = $existing
+            ? $existing->user
+            : ($this->getUser($providerUser) ?? $this->createsUser->create($provider, $providerUser));
+
+        $this->connectToProvider($user, $provider, $providerUser);
```

Comment:
This duplicates the query for finding `$existing`
Maybe should be combined with `createsUser`